src/pygeoprocessing/routing/routing.pyx

```diff
@@ -182,6 +183,8 @@ cdef cppclass GreaterPixel nogil:
         return 0
 
 cdef int _is_close(double x, double y, double abs_delta, double rel_delta):
+    if isnan(x) and isnan(y):

```

Comment:
Is it possible that `x` or `y` here is `nan` but not both? If so, does it still make sense to continue?

Reply:
Yes, it is absolutely possible that only one of these values may be `nan`!  Fortunately, the math of `abs(x-y) <= (abs_delta+rel_delta*abs(y))` works as expected (I'm using python here, but C and C++ both have equivalent results)

```python
def isclose(x, y, abs_delta, rel_delta):
     return abs(x-y) <= (abs_delta+rel_delta*abs(y))
```

```python
>>> isclose(1.5, float('nan'), 1e-8, 1e-5)
False
```

The problem this PR is trying to resolve is that comparing two `nan` values also is `False`:
```python
>>> isclose(float('nan'), float('nan'), 1e-8, 1e-5)
False
```

Reply:
Your comment also made me wonder if we might have a speedup if we did handle the case where only one of the values were `nan`, since we can avoid the arithmetic entirely if we do know that.

_Conclusion: I think what's in the PR is the best solution because it's easier to understand and the timings are so close anyways that I'm having a tough time getting any definitive timings._

I tested a couple different combinations, including the original (`base`) and found that on my laptop with 10 billion iterations, handling the exclusive-or case in a way that made sense took _way_ more time than I thought it would, doubling the function's apparent runtime.

It's worth noting that any improvements in times are only in the hundredsths or thousandths of a second faster over 10 billion runs, so averaged over 10 billion runs we're talking about differences of trillionths of a second per function call.  This is a level of timing precision that is unreliable, so I'm getting conflicting results over repeated runs.

So, that being said, the usually fastest version of `isclose` that I've found on my laptop a tossup between what's in the PR and a somewhat more convoluted implementation that sometimes reduces the number of times `isnan` is called:
```pyrex
cdef int _is_close(double x, double y, double abs_delta, double rel_delta):
    if isnan(x) and isnan(y):
        return 1
    return abs(x-y) <= (abs_delta+rel_delta*abs(y))

cdef int _is_close_xor_switch(double x, double y, double abs_delta, double rel_delta):
    if isnan(x):
        return isnan(y)
    return abs(x-y) <= (abs_delta+rel_delta*abs(y))
```

Here are one example of timing results:
```
2.0 elapsed time (base):  4.7931s 1
2.0 elapsed time (pr):    4.693s 1
2.0 elapsed time (xor1):  8.8043s 1
2.0 elapsed time (xorf):  8.7695s 1
2.0 elapsed time (xors):  4.6966s 1
2.0 elapsed time (xor):   8.7564s 1
nan elapsed time (base):  4.6996s 0
nan elapsed time (pr):    4.6963s 0
nan elapsed time (xor1):  7.9259s 0
nan elapsed time (xorf):  7.9238s 0
nan elapsed time (xors):  4.6906s 0
nan elapsed time (xor):   7.9262s 0
```

Here's the source code:
```pyrex
cdef int _is_close_base(double x, double y, double abs_delta, double rel_delta):
    return abs(x-y) <= (abs_delta+rel_delta*abs(y))

cdef int _is_close(double x, double y, double abs_delta, double rel_delta):
    if isnan(x) and isnan(y):
        return 1
    return abs(x-y) <= (abs_delta+rel_delta*abs(y))

cdef int _is_close_xor(double x, double y, double abs_delta, double rel_delta):
    if isnan(x) and isnan(y):
        return 1
    if isnan(x) ^ isnan(y):
        return 0
    return abs(x-y) <= (abs_delta+rel_delta*abs(y))

cdef int _is_close_xor_once(double x, double y, double abs_delta, double rel_delta):
    cdef int x_is_nan = isnan(x)
    cdef int y_is_nan = isnan(y)
    if x_is_nan and y_is_nan:
        return 1
    if x_is_nan ^ y_is_nan:
        return 0
    return abs(x-y) <= (abs_delta+rel_delta*abs(y))

cdef int _is_close_xor_first(double x, double y, double abs_delta, double rel_delta):
    cdef int x_is_nan = isnan(x)
    cdef int y_is_nan = isnan(y)
    if x_is_nan ^ y_is_nan:
        return 0
    if x_is_nan & y_is_nan:
        return 1
    return abs(x-y) <= (abs_delta+rel_delta*abs(y))

cdef int _is_close_xor_switch(double x, double y, double abs_delta, double rel_delta):
    if isnan(x):
        return isnan(y)
    else:
        return abs(x-y) <= (abs_delta+rel_delta*abs(y))

# re-run this once I have an internet connection.
def test_isclose():
    cdef unsigned long n = 10000000000  # 10 billion iterations
    cdef unsigned long i
    cdef unsigned long sum_ = 0
    cdef double y_value
    for y_value in [2, float('nan')]:
        start_time = time.time()
        sum_ = 0
        for i in range(n):
            sum_ += _is_close_base(i, y_value, 1e-8, 1e-5)
        print(f'{y_value} elapsed time (base):  {round(time.time() - start_time, 4)}s {sum_}')

        start_time = time.time()
        sum_ = 0
        for i in range(n):
            sum_ += _is_close(i, y_value, 1e-8, 1e-5)
        print(f'{y_value} elapsed time (pr):    {round(time.time() - start_time, 4)}s {sum_}')

        start_time = time.time()
        sum_ = 0
        for i in range(n):
            sum_ += _is_close_xor_once(i, y_value, 1e-8, 1e-5)
        print(f'{y_value} elapsed time (xor1):  {round(time.time() - start_time, 4)}s {sum_}')

        start_time = time.time()
        sum_ = 0
        for i in range(n):
            sum_ += _is_close_xor_first(i, y_value, 1e-8, 1e-5)
        print(f'{y_value} elapsed time (xorf):  {round(time.time() - start_time, 4)}s {sum_}')

        start_time = time.time()
        sum_ = 0
        for i in range(n):
            sum_ += _is_close_xor_switch(i, y_value, 1e-8, 1e-5)
        print(f'{y_value} elapsed time (xors):  {round(time.time() - start_time, 4)}s {sum_}')

        start_time = time.time()
        sum_ = 0
        for i in range(n):
            sum_ += _is_close_xor(i, y_value, 1e-8, 1e-5)
        print(f'{y_value} elapsed time (xor):   {round(time.time() - start_time, 4)}s {sum_}')

```